R/unlockREDCap.R

```diff
@@ -15,15 +14,120 @@
+         grepl("Couldn't connect to server", e))
+        stop("Unable to connect to url '",url,"'. ", e$message)
+        
+      if(grepl("403", e)) return(NULL)
+      
+      stop(e)
+    }
+  )
+}
 
- #############################################################################
-## Check if key is in package environment, aka memory
-.key_saved <- function(envir, key)
+  #############################################################################
+ ## unlock via YAML override if it exists
+##
+.unlockYamlOverride <- function(connections, url, ...)
 {
-  exists(key, envir=envir, inherits=FALSE) &&
-    !is.null(envir[[key]])                   &&
-    !is.na(envir[[key]])                     &&
-    !envir[[key]]==''
+  config_file <- file.path("..", paste0(basename(getwd()),".yml"))
+  
+  if(!file.exists(config_file)) return(list())
+  
+  config <- read_yaml(config_file)
+  if(is.null(config$redcapAPI)) stop(paste0("Config file '",config_file,"' does not contain required 'redcapAPI' entry"))
+  config <- config$redcapAPI
+  if(is.null(config$keys))      stop(paste0("Config file '",config_file,"' does not contain required 'keys' entry under the 'redcapAPI' entry"))
+  keys   <- config$keys
+  if(!is.null(config$args$url))  url <- config$args$url # Override from yml if available
+  config$args$url <- NULL
+  args   <- c(config$args, url = url, list(...))
+    
+  dest <- lapply(seq_along(connections), function(i) 
+  {
+    args$key  <- keys[connections[i]]
+    
+    if(is.null(args$key)) stop(paste0("Config file '", config_file, "' does not have API_KEY for '",connections[i],"' under redcapAPI: keys: specified."))
+    
+    do.call(.connectAndCheck, args)
+  })
+  names(dest) <- if(is.null(names(connections))) connections else names(connections)
+  
+  return(dest)
+}
+ 
+  #############################################################################
+ ## unlock keyring
+##
+.unlockKeyring <- function(keyring, passwordFUN)
+{
+  state <- keyring_list()
+  state <- state[state$keyring==keyring,]
+  
+  # If so, does it exist?
+  if(nrow(state) == 1) # Exists => UNLOCK
+  {
+    locked <- state$locked
+    # Is it locked
+    msg      <- paste0("Please enter password to unlock API keyring '",keyring, "'.")
+    while(locked)
+    {
+      password <- Sys.getenv("REDCAPAPI_PW")
+      stored   <- !is.null(password) && password != ''
+      if(!stored) password <- passwordFUN(msg)
+      if(is.null(password) || password == '') stop(paste0("User aborted keyring '",keyring, "' unlock."))
+      
+      tryCatch(
+        {
+          keyring_unlock(keyring, password)
+          Sys.setenv(REDCAPAPI_PW=password)
+          # Hacked work around for RStudio starting new session for everything
+          if(requireNamespace("rstudioapi", quietly = TRUE) &&
+            rstudioapi::isAvailable(child_ok=TRUE))
+            rstudioapi::sendToConsole(paste0("Sys.setenv(REDCAPAPI_PW='", password, "')"), execute = TRUE, echo=FALSE)
```

Comment:
```suggestion
            rstudioapi::sendToConsole(sprintf("Sys.setenv(REDCAPAPI_PW='%s')", password), execute = TRUE, echo=FALSE)
```

Reply:
very minor change; I prefer `sprintf` when more than one comma in `paste` command

Reply:
Done.